### ISSUE TITLE
long initialization overhead for GPU TEE cuda init

### ISSUE THREAD

**Issue by blossomin** [NVIDIA/nvtrust#113]
here is my test program, it simply called the cuInit(0);. but this function  call takes **cuInit time: 6557046 us**. this call should be about 500ms for a normal GPU. @rnertney , do you know where the overhead comes from for this heavy initialization?
`#include <iostream>
#include <cuda.h>
#include <chrono>

int main() {
    int gpu_count = 0;
    CUresult status;

    auto start = std::chrono::high_resolution_clock::now();
    status = cuInit(0);
    auto end = std::chrono::high_resolution_clock::now();
    auto duration_cuInit = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();

    if (status != CUDA_SUCCESS) {
        std::cerr << "cuInit failed: " << status << std::endl;
        return 1;
    }
    std::cout << "cuInit time: " << duration_cuInit << " us" << std::endl;
    return 0;
}`

**Comment by blossomin** [comment]
@thisiskarthikj @steven-bellock

**Comment by blossomin** [comment]
Thanks @steven-bellock , setting nvidia-persistenced to persistence works.

But I have a followup question: in the documents(https://docs.nvidia.com/cc-deployment-guide-tdx.pdf),
this section "Setting up the NVIDIA Driver to be in Persistence Mode" shows:
"However, in the SecureAI modes, this
behavior causes the shared-secret and the shared keys that were established during the
setup SPDM phase of the driver to be destroyed. Due to security concerns, the GPU will
not allow the restart of an SPDM session establishment without an FLR, which resets and
scrubs the GPU." 

if we don't set nvidia-persistenced to persistence, one SPDM session exits, and the key should be destroyed. And can relaunch another SPDM session without resetting the GPU? For example, can we execute the example CUDA program(cuInit) twice?

**Comment by steven-bellock** [comment]
>And can relaunch another SPDM session without resetting the GPU? For example, can we execute the example CUDA program(cuInit) twice?

No. `nvidia-persistenced` is required for confidential compute.

**Comment by blossomin** [comment]
Hi @steven-bellock , I have another question about concurrency in cuda with GPU TEE. 

Here is my new test program:

```
import os
import sys
import time

import torch
import time
t01 = time.time()
# Move the model to the appropriate device (CPU or GPU)
use_cuda = torch.cuda.is_available()
# use_cuda = True
t02 = time.time()
# warmup datamovement
tensor1 = torch.ones(1).cuda()
t03 = time.time()

print(f"pid: {os.getpid()}, use_cuda: {use_cuda}, use_cuda:{t02 - t01:.4f}s, warmup data move: {t03 - t02:.4f}s")
```

bash scripts:
```
for times in 1 10
do 
    mkdir results-$times
    for ((i=1; i<=times; i++))
    do
        python worker2.py > results-$times/output-$i.log 2>&1 &
    done
    sleep 60
done
```

the results are here:
for single test execution:
`pid: 9530, use_cuda: True, use_cuda:0.1078s, warmup data move: 1.4851s`

one sample of 10 concurrent test executions:
`pid: 9582, use_cuda: True, use_cuda:1.0719s, warmup data move: 14.9063s`

is this expected behavior?

**Comment by steven-bellock** [comment]
>is this expected behavior?

Somewhat. H2D/D2H transfers are hampered by having to encrypt and decrypt via software (OpenSSL) on the CPU. This can be sped up a little by using the newish AVX extensions and latest OpenSSL; see https://github.com/Azure/az-cgpu-onboarding/blob/main/docs/OpenSSL-Details.md. Also, encryption/decryption requires some serialization as there are IV counters that must be kept in-sync.

@moconnor725 might be able to provide more detail.

**Comment by moconnor725** [comment]
OpenSSL 3.1 and later can utilize AVX512 instructions which can improve copy bandwidth.  We've seen significant improvements using the [NVBandwidth  ](https://github.com/nvidia/nvbandwidth) when OpenSSL 3.1 is used.

**Comment by amshinde** [comment]
@blossomin I believe this is resolved. Closing this, please feel free to reopen if further assistance is needed.
